File: packages/identity-core/src/credential/bbs/bbs-credential.ts
import type { AgentContext } from '@credo-ts/core'
import { DidRepository, W3cCredentialsModuleConfig } from '@credo-ts/core'
import {
  BbsBlsSignature2020,
  BbsBlsSignatureProof2020,
  deriveProof as mattrDeriveProof,
} from '@mattrglobal/jsonld-signatures-bbs'
import { getBls12381G2KeyPair } from '../../kms/bbs-kms'
import {
  BBS_PROOF_TYPE_DERIVED,
} from './constants'
import { WEB_DID_KEY_BBS_LDP_FRAGMENT } from '../../did/registrar/web.registrar'

export type BbsDocumentLoader = (url: string) => Promise<{
  contextUrl: string | null
  documentUrl: string
  document: unknown
}>

export interface SignBbsCredentialOptions {
  /** Credencial JSON-LD sin proof. */
  credential: Record<string, unknown>
  /** DID URL de la verification method BLS (ej. `did:web:...#key-bbs-ldp`). */
  verificationMethod: string
  /** keyId KMS de la clave BLS del issuer. */
  kmsKeyId: string
  /**
   * Loader JSON-LD (contexts + DID Document).
   * Si se omite, se usa el de Credo `W3cCredentialsModuleConfig` vía `agentContext`.
   */
  documentLoader?: BbsDocumentLoader
}

export interface DeriveBbsProofOptions {
  /** VC firmada con BbsBlsSignature2020. */
  credential: Record<string, unknown>
  /** JSON-LD frame que indica qué claims revelar. */
  revealDocument: Record<string, unknown>
  /** Nonce / challenge del request-presentation. */
  nonce?: string
  documentLoader?: BbsDocumentLoader
}

/**
 * Adapta el documentLoader de Credo al contrato que espera jsonld-signatures / MATTR.
 */
export function resolveBbsDocumentLoader(
  agentContext: AgentContext,
  override?: BbsDocumentLoader
): BbsDocumentLoader {
  if (override) return override
  const moduleConfig = agentContext.dependencyManager.resolve(W3cCredentialsModuleConfig)
  const credoLoader = moduleConfig.documentLoader(agentContext)
  return async (url) => {
    const loaded = await credoLoader(url)
    return {
      contextUrl: loaded.contextUrl ?? null,
      documentUrl: loaded.documentUrl,
      document: loaded.document,
    }
  }
}

/**
 * Firma una credencial W3C JSON-LD con `BbsBlsSignature2020` usando la clave BLS del KMS.
 *
 * Credo 0.7 no modela BLS en `PublicJwk`; esta función es el camino canónico Quark (QUARK-990).
 * El material BLS vive en Postgres o en Vault KV según el driver Nest.
 */
export async function signBbsCredential(
  agentContext: AgentContext,
  options: SignBbsCredentialOptions
): Promise<Record<string, unknown>> {
  const keyPair = await getBls12381G2KeyPair(agentContext, options.kmsKeyId)
  const suite = new BbsBlsSignature2020({
    key: keyPair,
    verificationMethod: options.verificationMethod,
  })

  const documentLoader = resolveBbsDocumentLoader(agentContext, options.documentLoader)
  const jsigs = await import('jsonld-signatures')
  const signed = await jsigs.sign(options.credential, {
    suite,
    purpose: new jsigs.purposes.AssertionProofPurpose(),
    documentLoader,
  })
  return signed as Record<string, unknown>
}

/**
 * Verifica una VC con proof `BbsBlsSignature2020` o `BbsBlsSignatureProof2020`.
 *
 * La clave pública se toma del proof / documentLoader (DID del issuer), no del KMS local.
 */
export async function verifyBbsCredential(
  credential: Record<string, unknown>,
  options?: {
    documentLoader?: BbsDocumentLoader
  }
): Promise<{ verified: boolean; error?: string }> {
  try {
    const proof = credential.proof as { type?: string } | undefined
    const proofType = proof?.type
    const SuiteClass =
      proofType === BBS_PROOF_TYPE_DERIVED ? BbsBlsSignatureProof2020 : BbsBlsSignature2020
    const jsigs = await import('jsonld-signatures')
    const result = await jsigs.verify(credential, {
      suite: new SuiteClass(),
      purpose: new jsigs.purposes.AssertionProofPurpose(),
      documentLoader: options?.documentLoader,
    })
    return { verified: !!result.verified, error: result.error?.message }
  } catch (err: unknown) {
    return { verified: false, error: err instanceof Error ? err.message : String(err) }
  }
}

/**
 * Deriva un proof de selective disclosure (`BbsBlsSignatureProof2020`) desde una VC BBS.
 */
export async function deriveBbsProof(
  options: DeriveBbsProofOptions
): Promise<Record<string, unknown>> {
  const derived = await mattrDeriveProof(options.credential, options.revealDocument, {
    suite: new BbsBlsSignatureProof2020(),
    nonce: options.nonce ? Buffer.from(options.nonce) : undefined,
    documentLoader: options.documentLoader,
  })
  return derived as Record<string, unknown>
}

/**
 * Resuelve el kmsKeyId de `#key-bbs-ldp` para un DID creado en el wallet del agente.
 */
export async function resolveBbsKmsKeyId(
  agentContext: AgentContext,
  issuerDid: string
): Promise<{ verificationMethod: string; kmsKeyId: string }> {
  const didRepo = agentContext.dependencyManager.resolve(DidRepository)
  const record = await didRepo.findCreatedDid(agentContext, issuerDid)
  if (!record?.keys?.length) {
    throw new Error(`DID ${issuerDid} not found or has no keys in wallet`)
  }
  const bbsKey = record.keys.find((k) => k.didDocumentRelativeKeyId === WEB_DID_KEY_BBS_LDP_FRAGMENT)
  if (!bbsKey) {
    throw new Error(
      `Missing verification method ${WEB_DID_KEY_BBS_LDP_FRAGMENT} for ${issuerDid}. Create DID with addBbsKey.`
    )
  }
  return {
    verificationMethod: `${issuerDid}${WEB_DID_KEY_BBS_LDP_FRAGMENT}`,
    kmsKeyId: bbsKey.kmsKeyId,
  }
}

File: packages/identity-core-dart/lib/src/credential/bbs/bbs_credential.dart
import 'bbs_crypto.dart';
import 'constants.dart';
import 'reveal_frame.dart';

export 'bbs_crypto.dart';
export 'constants.dart';
export 'reveal_frame.dart';

/// Opciones para [deriveBbsProof].
class DeriveBbsProofOptions {
  const DeriveBbsProofOptions({
    required this.credential,
    required this.revealDocument,
    this.nonce,
    this.issuerDidDocument,
    this.crypto,
  });

  final Map<String, dynamic> credential;
  final Map<String, dynamic> revealDocument;
  final String? nonce;
  final Map<String, dynamic>? issuerDidDocument;
  final BbsCryptoBackend? crypto;
}

/// Deriva `BbsBlsSignatureProof2020` (selective disclosure) desde una VC BBS.
Future<Map<String, dynamic>> deriveBbsProof(DeriveBbsProofOptions options) {
  final proof = options.credential['proof'];
  final proofType = proof is Map ? proof['type'] as String? : null;
  if (proofType != kBbsProofType) {
    throw ArgumentError(
      'deriveBbsProof espera proof.type=$kBbsProofType, recibido=$proofType',
    );
  }
  return (options.crypto ?? bbsCrypto).deriveProof(
    credential: options.credential,
    revealDocument: options.revealDocument,
    nonce: options.nonce,
    issuerDidDocument: options.issuerDidDocument,
  );
}

/// Verifica VC con proof BBS (Signature2020 o Proof2020).
Future<({bool verified, String? error})> verifyBbsCredential(
  Map<String, dynamic> credential, {
  Map<String, dynamic>? issuerDidDocument,
  BbsCryptoBackend? crypto,
}) {
  final proof = credential['proof'];
  final proofType = proof is Map ? proof['type'] as String? : null;
  if (!isBbsProofType(proofType)) {
    return Future.value((
      verified: false,
      error: 'proof.type no es BBS ($proofType)',
    ));
  }
  return (crypto ?? bbsCrypto).verifyCredential(
    credential: credential,
    issuerDidDocument: issuerDidDocument,
  );
}

/// Si la VC es `BbsBlsSignature2020`, deriva según paths del PD; si no, la deja igual.
Future<Map<String, dynamic>> maybeDeriveBbsCredentialForPex({
  required Map<String, dynamic> credential,
  required Map<String, dynamic> presentationDefinition,
  String? nonce,
  Map<String, dynamic>? issuerDidDocument,
  BbsCryptoBackend? crypto,
}) async {
  final proof = credential['proof'];
  final proofType = proof is Map ? proof['type'] as String? : null;
  if (proofType != kBbsProofType) {
    return Map<String, dynamic>.from(credential);
  }

  final paths =
      extractRevealPathsFromPresentationDefinition(presentationDefinition);
  if (paths.isEmpty) {
    return Map<String, dynamic>.from(credential);
  }

  final frame = buildRevealFrame(credential, paths);
  final derived = await deriveBbsProof(
    DeriveBbsProofOptions(
      credential: credential,
      revealDocument: frame,
      nonce: nonce,
      issuerDidDocument: issuerDidDocument,
      crypto: crypto,
    ),
  );

  // Holder binding: restaurar credentialSubject.id si el frame no lo preservó.
  final originalSubject = credential['credentialSubject'];
  final originalId =
      originalSubject is Map ? originalSubject['id'] as String? : null;
  final derivedSubject = derived['credentialSubject'];
  if (originalId != null &&
      derivedSubject is Map &&
      derivedSubject['id'] is! String) {
    derived['credentialSubject'] = {
      ...Map<String, dynamic>.from(derivedSubject),
      'id': originalId,
    };
  }

  return derived;
}
